niceday-api/index.test.js

```diff
@@ -1,5 +1,80 @@
-describe('Temporary Mock Test Suite', () => {
-  it('Mock Test Case', () => {
-    expect(true).toEqual(true);
+require('isomorphic-fetch');
+
+const NICEDAY_TEST_SERVERPORT = 8080;
+const NICEDAY_TEST_USER_ID = 38527;
+const MOCK_USER_DATA = {
+  id: NICEDAY_TEST_USER_ID,
+  userProfile: {
+    firstName: 'Mr Mock',
+  },
+};
+
+// Contains all tests which require a mocked Senseserver
+describe('Tests on niceday-api server using mocked goalie-js', () => {
+  let server;
+
+  beforeAll(() => {
```

Comment:
I am curious if we want to do this `beforeAll` always in all future tests. If I think out loud, the SenseServer and Authentication need to be mocked always, but the `getContact` method only needs to be mocked for the `get userdata` test. If you want to implement it that way you will definitely get the 'jest-guru' award, otherwise you can leave it or create an issue for it.

Reply:
This is something I struggled a bit with. The problem is that you need to have mocked the `goaliejs` module _before_ the server is started - that's why
```js
const { createNicedayApiServer } = require('./index');
```
is in the `beforeEach()` and not at the start of the file.

I've tried solutions along the lines of e.g.
```js
    goaliejs = require('@sense-os/goalie-js');
...
...
...
    goaliejs.SenseServer = jest.fn( () => ({
      Alpha: undefined,
    }));

    goaliejs.SenseNetwork = jest.fn().mockImplementation(() => ({
      getContact: () => new Promise((resolve) => {
        resolve(MOCK_USER_DATA);
      }),
    }));

    goaliejs.Authentication = jest.fn().mockImplementation(() => ({
      login: () => new Promise((resolve) => {
        console.debug('Mocking successful authentication');
        const mockAuthResponse = {
          token: 'mocktoken',
          user: {
            id: 12345,
          },
        };
        resolve(mockAuthResponse);
      }),
    }));
```

So that you could swap in mocked submodules like `SenseNetwork` later (in the specific `userdata` test, for example) but it doesn't work because this is only changing our reference to goaliejs, not the one that `index.js` will fetch. So you have to use `jest.mock()`. In this vein, I tried:
```js
    jest.mock('@sense-os/goalie-js', () => ({
      SenseNetwork: jest.fn().mockImplementation(() => ({
        getContact: () => new Promise((resolve) => {
          resolve(MOCK_USER_DATA);
        }),
      })),
    }));
```
in the userdata test (obviously without specifying `SenseNetwork` in the `beforeAll` mock) but then that fails because things like `service/UserDataService.js` actually get 'compiled' when `server = createNicedayApiServer();` is first called.

So I'm not sure there is a nice way to do this without mocking everything we want to in `goaliejs` before starting the server up. The alternative would we not using `beforeEach` etc at all and constructing a new mock within each test, but I think that will actually be more complex.

However, I definitely don't know enough jest/js to say with certainty that there's no clean solution.

Reply:
OK, thanks for the extensive answer. We can always return back to this discussion in case we do want it differently!